packages/services/api/src/modules/preflight-script/providers/preflight-script.provider.ts

```diff
@@ -0,0 +1,215 @@
+    console.log({ error });
+    return {
+      error: {
+        __typename: 'PreflightScriptError' as const,
+        message: error instanceof Error ? error.message : String(error),
+      },
+    };
+  }
+}
+
+@Injectable({
+  global: true,
+  scope: Scope.Operation,
+})
+export class PreflightScriptProvider {
+  private logger: Logger;
+
+  constructor(
+    logger: Logger,
+    private storage: Storage,
+    private session: Session,
+    private idTranslator: IdTranslator,
+    @Inject(PG_POOL_CONFIG) private pool: DatabasePool,
+  ) {
+    this.logger = logger.child({ source: 'PreflightScriptProvider' });
+  }
+
+  async getPreflightScript(targetSlug: string) {
+    const result = await this.pool.maybeOne(sql`/* getPreflightScript */
+      SELECT "id"
+           , "source_code"         as "sourceCode"
+           , "target_id"           as "targetId"
+           , "created_by_user_id"  as "createdByUserId"
+           , to_json("created_at") as "createdAt"
+           , to_json("updated_at") as "updatedAt"
+      FROM "document_preflight_scripts"
+      WHERE "target_id" = ${targetSlug}
+      `);
+
+    return result && PreflightScriptModel.parse(result);
+  }
+
+  async createPreflightScript(
+    selector: TargetSelectorInput,
+    { sourceCode }: CreatePreflightScriptInput,
+  ) {
+    const res = validateSourceCode(sourceCode);
+    if (res) return res;
+
+    const [organizationId, projectId, targetId] = await Promise.all([
+      this.idTranslator.translateOrganizationId(selector),
+      this.idTranslator.translateProjectId(selector),
+      this.idTranslator.translateTargetId(selector),
+    ]);
+
+    await this.session.assertPerformAction({
+      action: 'laboratory:createPreflightScript',
+      organizationId,
+      params: {
+        organizationId,
+        projectId,
+        targetId,
+      },
+    });
+
+    const target = await this.storage.getTarget({
+      organizationId,
+      projectId,
+      targetId,
+    });
+
+    const currentUser = await this.session.getViewer();
+
+    const result = await this.pool.maybeOne(sql`/* createPreflightScript */
+      INSERT INTO "document_preflight_scripts" ( "source_code"
+                                               , "target_id"
+                                               , "created_by_user_id")
+      VALUES (${sourceCode},
+              ${targetId},
+              ${currentUser.id})
+      RETURNING
+          "id"
+          , "source_code" as "sourceCode"
+          , "target_id" as "targetId"
+          , "created_by_user_id" as "createdByUserId"
+          , to_json("created_at") as "createdAt"
+          , to_json("updated_at") as "updatedAt"
+      `);
+    const { data: preflightScript, error } = PreflightScriptModel.safeParse(result);
+
+    if (error) {
+      const { message } = fromZodError(error);
+      return {
+        error: {
+          __typename: 'PreflightScriptError' as const,
+          message,
+        },
+      };
+    }
```

Comment:
move this before you are doing a DB insert. Actually now if I think about it on what I was saying about `zod.superRefine` will make much more sense

Reply:
thank you, super refine is nice, encapsulate error logic for source code https://github.com/graphql-hive/console/pull/5564#discussion_r1848045797